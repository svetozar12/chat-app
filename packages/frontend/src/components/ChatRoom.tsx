import React, { useState, useEffect, ChangeEvent } from "react";
import { NextPage } from "next";
import { io, Socket } from "socket.io-client";
import { useDispatch, useSelector } from "react-redux";
import { InitialState, InitialState2 } from "../redux/state";
import { MdSend } from "react-icons/md";
import axios from "axios";
import RenderChat from "../components/RenderChat";
import timeStamp from "../utils/timeStamp";
import { requestUrl } from "../utils/hostUrl_requestUrl";
import { useRouter } from "next/router";
import ChatHeader from "./ChatHeader";

interface IHome {
  cookie: any;
  chatId: any;
}

interface IPropsState {
  name?: string;
  message?: string;
  time?: string | number;
}

interface IchatInstance {
  sender: string;
  message: string;
  createdAt: string;
}

const ChatRoom: NextPage<IHome> = ({ cookie, chatId }) => {
  const route = useRouter();
  const statess = useSelector(
    (state: { setReducer: InitialState2 }) => state.setReducer,
  );

  const cookieName = cookie.get("name");
  const dispatch = useDispatch();
  const [chat, setChat] = useState<IchatInstance[]>([]);
  const [socketRef, setSocketRef] = useState<Socket | null>(null);
  const [state, setState] = useState<IPropsState>({
    name: cookie.get("name"),
    message: "",
    time: "",
  });

  const updateChat = (param: IchatInstance[]) => {
    setChat((prev) => [...prev, ...param]);
  };

  const getRecentMessages = async () => {
    try {
      const res = await axios.get(
        `${requestUrl}/messages/${chatId}?page_number=1&page_size=10`,
      );
      const data = res.data.reversedArr;
      console.log(data);

      updateChat(data);
      return true;
    } catch (error) {
      return false;
    }
  };

  useEffect(() => {
    setChat([]);
    getRecentMessages();
  }, [route.asPath]);

  useEffect(() => {
    if (!cookie.get("token") && !cookie.get("refresh_token"))
      dispatch({ type: "SIGN_OUT" });
    const socketConnect: Socket = io("http://localhost:4000");
    socketConnect.on("message", ({ messages }) => {
      dispatch({
        type: "MESSAGE_SEND",
        payload: { sender: cookieName, message: state.message },
      });
      updateChat(messages);
    });
    socketConnect.emit("joined_chat_room", { user: cookieName });
    setSocketRef(socketConnect);
    return () => {
      socketRef && socketRef.disconnect();
    };
  }, []);

  const onTextChange = (e: ChangeEvent<HTMLTextAreaElement>): void => {
    setState({ ...state, [e.target.name]: e.target.value });
  };

  const saveMessage = async () => {
    try {
      const res = await axios.post(`${requestUrl}/messages/${chatId}`, {
        sender: cookieName,
        message: state.message,
      });

      dispatch({
        type: "MESSAGE_SEND",
        payload: { sender: cookieName, message: state.message },
      });

      return true;
    } catch (error) {
      return false;
    }
  };

  const scrollHandler = async (e: React.UIEvent<HTMLElement>) => {
    try {
      if (e.currentTarget.scrollTop === 0) {
        dispatch({
          type: "INCREMENT_PAGE_NUMBER",
          payload: statess.pageNumber,
        });
        const res = await axios.get(
          `${requestUrl}/messages/${chatId}?page_number=${statess.pageNumber}&page_size=10`,
        );
        const data = res.data.reversedArr;

        setChat((prev) => [...data, ...prev]);
      }
      return true;
    } catch (error) {
      return false;
    }
  };

  const onMessageSubmit = async (
    e: React.FormEvent<HTMLFormElement> | React.MouseEvent<SVGElement>,
  ) => {
    e.preventDefault();
    if (state.message) {
      const { name, message, time } = state;

      await saveMessage();
      socketRef?.emit("message", {
        chatInstance: chatId,
        sender: cookieName,
        message,
        time,
      });
      setState({ name, message: "" });
    }
  };

  const handleKeyPress = (e: any) => {
    console.log(e.key);

    if (e.key === "Enter") {
      onMessageSubmit(e);
    }
  };

  return (
    <div
      style={{
        justifyContent: "center",
        alignItems: "center",
        width: "100%",
        height: "100vh",
        padding: "0",
      }}
      className="container  chat_home"
    >
      {socketRef && statess.toggleCreateGroup && (
        <ChatHeader socketRef={socketRef} cookieName={cookie.get("name")} />
      )}

      <div onScroll={scrollHandler} className="container_chat">
        {chat.map((item, index) => {
          const { sender, message, createdAt } = item;
          const time_stamp = timeStamp(createdAt);
          return (
            <li style={{ listStyle: "none" }} key={index}>
              <RenderChat
                key={index}
                chatId={chatId}
                cookie={cookie.get("name")}
                sender={sender}
                time_stamp={time_stamp}
                message={message}
              />
            </li>
          );
        })}
      </div>

      <form
        onSubmit={onMessageSubmit}
        className="message_form"
        style={{ width: "100%" }}
      >
        <div className="message_input_container">
          <textarea
            rows={1}
            name="message"
            onChange={(e) => onTextChange(e)}
            style={{ border: "none", resize: "none" }}
            placeholder="Your Message "
            value={state.message}
            onKeyPress={(e: any) => handleKeyPress(e)}
          ></textarea>
          <MdSend type="submit" onClick={onMessageSubmit} />
        </div>
      </form>
    </div>
  );
};

export default ChatRoom;
