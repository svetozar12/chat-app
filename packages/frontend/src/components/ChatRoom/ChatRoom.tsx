import { NextPage } from "next";
import { useRouter } from "next/router";
import React, { useEffect } from "react";
// utils
import timeStamp from "../../utils/timeStamp";
import { constants } from "../../constants";
// components
import RenderChat from "./RenderChat";
import ChatHeader from "../ChatHeader";
import ChatRoomForm from "./ChatRoomForm";
import SkelletonUserMessages from "../Loading/SkelletonUserMessages";
// hooks
import { useDispatch, useSelector } from "react-redux";
import { useCookie } from "next-cookie";
// services
import api_helper from "../../services/graphql/api_helper";
import { InitialStateMessage } from "../../services/redux/reducer/messageReducer/state";
import { IInitialSet } from "../../services/redux/reducer/setReducer/state";
import { VStack } from "@chakra-ui/react";
import { useAuth } from "../../utils/SessionProvider";

interface IHome {
  chatId: any;
}

export interface IchatInstance {
  _id: string;
  sender: string;
  message: string;
  createdAt: string;
}

const ChatRoom: NextPage<IHome> = ({ chatId }) => {
  const route = useRouter();
  const messageState = useSelector((state: { messageReducer: InitialStateMessage }) => state.messageReducer);
  const setState = useSelector((state: { setReducer: IInitialSet }) => state.setReducer);
  const cookie = useCookie();

  const user = useAuth();
  const user_id = cookie.get("id") as string;
  const token = cookie.get("token") as string;
  const dispatch = useDispatch();
  const containerRef = React.useRef<null | HTMLDivElement>(null);

  const getRecentMessages = async () => {
    try {
      const res = await api_helper.message.getAll({ user_id, chat_id: chatId, token, query: { page_size: 10, page_number: 1 } });

      res.forEach((element) => {
        dispatch({ type: "MESSAGES", payload: element });
      });

      return true;
    } catch (error) {
      return false;
    }
  };

  useEffect(() => {
    dispatch({ type: "SET_IS_MATCH", payload: false });
    if (location.href === constants.HOST_URL + "/" + chatId) dispatch({ type: "SET_IS_MATCH", payload: true });
    dispatch({ type: "RESET_MESSAGES" });
    getRecentMessages();
  }, [route.asPath]);

  const scrollHandler = async (e: React.UIEvent<HTMLElement>) => {
    try {
      if (e.currentTarget.scrollTop === 0) {
        dispatch({
          type: "INCREMENT_PAGE_NUMBER",
          payload: setState.pageNumber,
        });
        const res = await api_helper.message.getAll({
          user_id,
          chat_id: chatId,
          token,
          query: { page_size: 10, page_number: setState.pageNumber },
        });
        const data = res.reversedArr;

        data.forEach((element) => {
          dispatch({ type: "PAGGINATION_MESSAGES", payload: element });
        });
      }
      return true;
    } catch (error) {
      return false;
    }
  };

  const scrollToBottom = () => {
    const parent = containerRef.current;
    parent?.scrollTo(0, parent.scrollHeight);
  };

  React.useEffect(() => {
    scrollToBottom();
  }, [messageState.messages]);

  return (
    <VStack w="full" h="100vh">
      {setState.toggleCreateGroup && <ChatHeader />}

      {user ? (
        <VStack
          w="full"
          mt={{ base: "2rem", lg: "-0.5rem !important" }}
          h="full"
          p="1rem"
          overflow="auto"
          bg="main_white"
          ref={containerRef}
          onScroll={scrollHandler}
        >
          {messageState.messages.map((item, index) => {
            const { sender, message, createdAt } = item;
            const time_stamp = timeStamp(createdAt);

            return <RenderChat key={index} chatId={chatId} id={item._id} sender={sender} time_stamp={time_stamp} message={message} />;
          })}
        </VStack>
      ) : (
        <SkelletonUserMessages />
      )}

      <ChatRoomForm chatId={chatId} />
    </VStack>
  );
};

export default ChatRoom;
