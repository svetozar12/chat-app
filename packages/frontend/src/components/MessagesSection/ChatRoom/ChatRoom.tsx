import { useRouter } from 'next/router';
import React, { useEffect } from 'react';
// utils
import { connect } from 'react-redux';
import { useCookie } from 'next-cookie';
import { VStack } from '@chakra-ui/react';
import timeStamp from '../../../utils/timeStamp';
import { constants } from '../../../constants/index';
// components
import RenderChat from './RenderChat';
import ChatHeader from './ChatHeader';
import ChatRoomForm from './ChatRoomForm';
import SkelletonUserMessages from '../../Loading/SkelletonUserMessages';
// hooks
// services
import { gqlSdk } from '@chat-app/sdk';
import { useAuth } from '../../../utils/SessionProvider';
import useThemeColors from '../../../hooks/useThemeColors';
import { STATE } from 'services/redux/reducer';
import { bindActionCreators, Dispatch } from 'redux';
import {
  incrementPaginationNumberAction,
  resetMessagesAction,
  setMessagesAction,
  setPaginatedMessagesAction,
} from 'services/redux/reducer/messages/actions';
import { IMessage } from 'services/redux/reducer/messages/state';
import { IToggle } from 'services/redux/reducer/toggles/state';
import { toggleIsMatch } from 'services/redux/reducer/toggles/actions';

interface IChatRoom {
  message: IMessage;
  toggle: IToggle;
  chatId: string;
  incrementPagination: typeof incrementPaginationNumberAction;
  setMessages: typeof setMessagesAction;
  setPaginatedMessages: typeof setPaginatedMessagesAction;
  resetMessages: typeof resetMessagesAction;
  toggleIsMatch: typeof toggleIsMatch;
}

export interface IchatInstance {
  _id: string;
  sender: string;
  message: string;
  createdAt: string;
}

function ChatRoom(props: IChatRoom) {
  const { chatId, incrementPagination, setPaginatedMessages, setMessages, resetMessages, message, toggle } = props;
  const { messagePageNumber, messages, show } = message;
  const route = useRouter();
  const cookie = useCookie();
  const {
    colors: { chatBg },
  } = useThemeColors();
  const user = useAuth();
  const userId: string = cookie.get('id');
  const token: string = cookie.get('token');
  const containerRef = React.useRef<null | HTMLDivElement>(null);

  const getRecentMessages = async () => {
    try {
      const res = await gqlSdk.message.getAll({ userId, chatId, token, query: { page_size: 10, page_number: 1 } });

      res.forEach((element: Record<string, any>) => {
        setMessages(element);
      });

      return true;
    } catch (error) {
      return false;
    }
  };

  useEffect(() => {
    toggleIsMatch(false);
    if (location.href === `${constants.HOST_URL}/${chatId}`) toggleIsMatch(true);
    resetMessages();
    getRecentMessages();
  }, [route.asPath]);

  const scrollHandler = async (e: React.UIEvent<HTMLElement>) => {
    try {
      if (e.currentTarget.scrollTop === 0) {
        incrementPagination(messagePageNumber);
        const res = await gqlSdk.message.getAll({
          userId,
          chatId,
          token,
          query: { page_size: 10, page_number: messagePageNumber },
        });
        const data = res.reversedArr;

        data.forEach((element: Record<string, any>) => {
          setPaginatedMessages(element);
        });
      }
      return true;
    } catch (error) {
      return false;
    }
  };

  const scrollToBottom = () => {
    const parent = containerRef.current;
    parent?.scrollTo(0, parent.scrollHeight);
  };

  React.useEffect(() => {
    scrollToBottom();
  }, [messages]);

  return (
    <VStack w="full" h="100vh">
      {toggle.toggleCreateGroupModal && <ChatHeader />}

      {user ? (
        <VStack
          w="full"
          mt={{ base: '2rem', lg: '-0.5rem !important' }}
          h="full"
          p="1rem"
          overflow="auto"
          bg={chatBg}
          ref={containerRef}
          onScroll={scrollHandler}
        >
          {messages.map((item, index) => {
            const { sender, message, createdAt } = item;
            const TimeStamp = timeStamp(createdAt);

            return <RenderChat key={index} chatId={chatId} id={item._id} sender={sender} timeStamp={TimeStamp} recievedMessage={message} />;
          })}
        </VStack>
      ) : (
        <SkelletonUserMessages />
      )}

      <ChatRoomForm chatId={chatId} />
    </VStack>
  );
}

const mapStateToProps = (state: STATE) => ({
  message: state.messages,
  toggle: state.toggle,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  incrementPagination: bindActionCreators(incrementPaginationNumberAction, dispatch),
  setMessages: bindActionCreators(setMessagesAction, dispatch),
  setPaginatedMessages: bindActionCreators(setPaginatedMessagesAction, dispatch),
  toggleIsMatch: bindActionCreators(toggleIsMatch, dispatch),
  resetMessages: bindActionCreators(resetMessagesAction, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(ChatRoom);
