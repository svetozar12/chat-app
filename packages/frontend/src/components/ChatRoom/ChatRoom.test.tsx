import ChatRoom from "../ChatRoom/ChatRoom";
import renderer from "react-test-renderer";
import { render, cleanup, RenderResult } from "@testing-library/react";
import { ReactTestRendererJSON } from "react-test-renderer";
import "@testing-library/jest-dom";

let component: ReactTestRendererJSON | ReactTestRendererJSON[] | null;
let container: RenderResult;
const socketRef: any = jest.fn();
const submit: any = jest.fn();

beforeEach(() => {
  component = renderer
    .create(<ChatRoom cookie={submit} chatId={"321312313"} />)
    .toJSON();

  container = render(<ChatRoom cookie={submit} chatId={"321312313"} />);
});

afterEach(cleanup);

describe("Render connected React-redux page", () => {
  it("should create snapshot for <LoginForm/>", () => {
    expect(component).toMatchSnapshot();
  });

  it("should render <Group_avatar/>", () => {
    const renderedComponent = container.getByRole("textarea");
    expect(renderedComponent).toBeInTheDocument();
  });
});
