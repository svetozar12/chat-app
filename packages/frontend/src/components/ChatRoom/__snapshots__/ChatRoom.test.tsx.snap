// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render connected React-redux page should create snapshot for <ChatRoom/> 1`] = `
<div
  className="css-1708wod"
>
  <div
    className="css-1pt27nv"
    onScroll={[Function]}
  />
  <form
    className="flex css-16nrlu7"
    onSubmit={[Function]}
  >
    <div
      className="css-1q6ynyn"
      onClick={[Function]}
    >
      <textarea
        className="css-1b140on"
        name="message"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Your Message "
        value=""
      />
      <div
        className="css-s85sp7"
      >
        <svg
          className="css-1ao2df9"
          fill="currentColor"
          height="1em"
          onClick={[Function]}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          type="submit"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
          />
        </svg>
      </div>
    </div>
  </form>
</div>
`;
